src/Database/Driver/SqlDialectTrait.php

```diff
@@ -143,9 +145,92 @@ protected function _expressionTranslators(): array
      */
     protected function _selectQueryTranslator(Query $query): Query
     {
+        $query = $this->_generateSelectExpressions($query);
+
         return $this->_transformDistinct($query);
     }
 
+    /**
+     * Generates expressions for selected fields whose mapped types
+     * implement `\Cake\Database\Type\SelectExpressionTypeInterface`.
+     *
+     * @param \Cake\Database\Query $query The query to modify.
+     * @return \Cake\Database\Query The modified query.
+     */
+    protected function _generateSelectExpressions(Query $query): Query
+    {
+        $typeMap = $query->getSelectTypeMap()->toArray();
+        if (!$typeMap) {
+            return $query;
+        }
+
+        $isAutoQuotingEnabled = $this->isAutoQuotingEnabled();
+        $select = $query->clause('select');
+
+        foreach ($select as $alias => $field) {
+            // Only handle possible field names and identifier expressions,
+            // other complex expression are going to be assumed to already
+            // handle conversion required for the possible custom data type.
```

Comment:
That would be the case, yes. I considered a deep replace, but I came to the conclusion that it could easily interfere with other calculations/conversions.

I think most of the time when using custom expressions in the `SELECT` clause one would want the actual unaltered field, as opposed to what one would want to use for the queried results. Say for example I want to calculate the distance between two points, I'd do it like this:

```php
[
    'dist' => $query->func()->ST_Distance([
        $query->identifier('Points.pt_start'),
        $query->identifier('Points.pt_end'),
    ])
]
```

That would generate SQL like:

```sql
ST_Distance(Points.pt_start, Points.pt_end) AS dist
```

Now if the identifiers would be wrapped in `ST_AsGeoJSON()` function calls, which is what I would want for `toPHP()` conversion, then the query would look like this and break, as JSON data is invalid for the `ST_Distance()` function arguments:

```sql
ST_Distance(ST_AsGeoJSON(Points.pt_start), ST_AsGeoJSON(Points.pt_end)) AS dist
```

Reply:
In that situation the expression substitution wouldn't be good, but in arithmetic operations or aggregations you might want the translations. My concern is that we're providing an abstraction that isn't consistently applied and that we'll end up fielding issues and adding documentation complexity.

[SQLAlchemy approaches custom types](https://docs.sqlalchemy.org/en/13/core/custom_types.html#creating-new-types) in a similar but simpler way. Their system provides hooks for custom column definitions and hooks to cast data going in to queries and out of results. They don't offer what the select expression substitution though. I think that is the riskiest part of these changes and I think we might be better off if users had to explicitly use the requisite SQL functions in order to get that behavior.

Reply:
That's a fair point, I expected that would be a little controversial. I can see how it feels somewhat incomplete, and how it could come off as possibly unexpected behavior to people.

Looking at the SQL Alchemy docs, isn't **[`TypeEngine.column_expression`](https://docs.sqlalchemy.org/en/13/core/custom_types.html#applying-sql-level-bind-result-processing)** pretty much the equivalent to this? If it is, I'm wondering if they're maybe doing some clever stuff with existing expressions that could be leveraged here, but I'm not even sure if I understand the docs correctly, I have zero experience with SQL Alchemy.

Alternatively maybe the type could act as a processor for the relevant select list entries, ie it would have to traverse the possible expressions that the fields holds manually, that would still keep things close to the type. It's kinda similar to `beforeFind()` then though, however with the difference that it would receive the final selected fields, where as the select list might be completely empty in `beforeFind()`.

Reply:
So I've tried to gather a little more understanding of SQL Alchemy. `TypeEngine.column_expression` indeed seems to be the equivalent here, and it looks like there is actually some kind of decision making going on for expressions.

The following setup:

```python
from sqlalchemy import func
from sqlalchemy.types import UserDefinedType

class PointGeometry(UserDefinedType):
    def column_expression(self, col):
        return func.ST_AsGeoJSON(col, type_=self)

from sqlalchemy.ext.declarative import declarative_base
Base = declarative_base()

from sqlalchemy import Column, Integer, String
class Point(Base):
    __tablename__ = 'points'
   
    id = Column(Integer, primary_key = True)
    pt_start = Column(PointGeometry)
    pt_end = Column(PointGeometry)

points = Point.__table__

from sqlalchemy import select
stmt = select([
    points.c.pt_start,
    func.ST_AsText(points.c.pt_start),
    func.ST_Distance(points.c.pt_start, points.c.pt_end),
    func.custom(points.c.pt_start),
    func.custom(func.sum(points.c.pt_start)),
    func.sum(points.c.pt_start),
    func.max(points.c.pt_start),
    func.count(points.c.pt_start),
    func.count().over(partition_by=points.c.pt_start, order_by=points.c.pt_start),
])

print(stmt)
```

will generate this SQL:

```sql
SELECT
    ST_AsGeoJSON(points.pt_start) AS pt_start,
    ST_AsText(points.pt_start) AS "ST_AsText_1",
    ST_Distance(points.pt_start, points.pt_end) AS "ST_Distance_1",
    custom(points.pt_start) AS custom_1,
    custom(sum(points.pt_start)) AS custom_2,
    ST_AsGeoJSON(sum(points.pt_start)) AS sum_1,
    ST_AsGeoJSON(max(points.pt_start)) AS max_1,
    count(points.pt_start) AS count_1,
    count(*) OVER (PARTITION BY points.pt_start ORDER BY points.pt_start) AS anon_1
FROM points
```

As can be seen, it wraps plain fields in `ST_AsGeoJSON()`, for "unknown" functions like `ST_Distance()` it won't, it will use the plain fields, while for some known functions like `sum()` it will wrap the result in `ST_AsGeoJSON()`, while for other known ones like `count()` it won't. Of course using a point type with for example `sum()` doesn't make any sense, I just wanted to see what the generate SQL would look like.

I'm wondering, @markstory , do you think something like that could work? I guess the fact that the outer most expression is being wrapped instead of the innermost identifier could be a problem in certain situations, and similarly it could be a problem vice versa, someone would always be unhappy about either behavior 🤔

Reply:
> Looking at the SQL Alchemy docs, isn't TypeEngine.column_expression pretty much the equivalent to this? 

Yes it is. I didn't see that section of the docs. Having the type be able to transform the field into a new expression could work well. Would you want to only do that transform when the field is selected directly still?

Reply:
> Having the type be able to transform the field into a new expression could work well.

I'm not sure that I understand correctly what exactly you're referring to with "_transform the field into a new expression_"? The wrapping behavior that SQL Alchemy applies, or the idea of passing the field(s) to the type irrespective of what expression they might be wrapped in so that the type has full control over what happens with the field?

> Would you want to only do that transform when the field is selected directly still?

Currently fields are processed irrespective of whether they were selected explicitly or implicitly (via auto fields). I would've kept it that way.